metabase: fix HEAD builds

chrmoritz opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

During the [npm@5 upgrade in the node formula](https://github.com/Homebrew/homebrew-core/commit/ecb2d1e8551d80b5c5de2ac0d00dc071987eac01#diff-d3ef2414308e30df24a517ca9a1324eeR87) the location of npm was changed from `libexec/npm/bin` to just `libexec/bin` and homebrew-npm's path was [updated inside language/node](https://github.com/Homebrew/brew/commit/b6d81359f0a958d27770280900cfb55fbbd62d5e#diff-876a0ba7f64fe71b16f0c695b0e60c02R28). But this core formula, which is directly hardcoding the npm path, was overlooked. This PR fixes the npm path by using language/nodes `setup_npm_environment` method, so that the path to npm is only maintained in one location (inside language/node), which should hopefully prevent similar issues in the future.

Also it looks like HEAD builds are depending on `yarn` now.

chrmoritz commented on Formula/metabase.rb:
```diff
@@ -21,6 +21,8 @@ class Metabase < Formula
     if build.head?
       Language::Node.setup_npm_environment
+      # help yarn find npm's node-gyp binary inside superenv
+      ENV.prepend_path "PATH", "#{Formula["node"].opt_libexec}/lib/node_modules/npm/bin/node-gyp-bin"
```
Ideally we should either fix this inside the yarn formula somehow or we should create a language/node helper for it too, so that it's $PATH is only managed in one place for the exact same reason we use `setup_npm_environment` now.

And just the clarify: homebrew's `yarn` is able to compile native addons with `node-gyp` outside a formula completely fine. It's superenv which break yarn ability to find `node-gyp` when used in formulas a a build dependency.

DomT4 replied:
Is there any harm to simply chucking it into the `$PATH` unconditionally when `"language/node"` is `require`d?

chrmoritz replied:
No this could be done, but the underlying issue is that yarn auto-detects it's global installation prefix as the bin folder inside node's Cellar. If you for example globally install webpack with `yarn global add webpack`, yarn will put a symlink into `HOMEBREW_PREFIX/Cellar/node/8.2.1/bin/webpack`, which is imho harmful (nothing should manipulate another formulas cellar).

A fix would be to run `yarn config set prefix HOMEBREW_PREFIX`, but this can't be done in a post_install step because the sandbox will block write request to `~/.yarnrc` if not run with `--no-sandbox`.

chrmoritz replied:
Also setting the correct global installation prefix to `~/.yarnrc` would help homebrew users to finally be able to install stuff globally with yarn, but it won't help formulas using yarn as a build dependency, because brew manipulates $HOME during installation and yarn would look in the wrong place again.

yarn installing node-gyp inside a formula environment:
```
error An unexpected error occurred: "EPERM: operation not permitted, symlink '../../../../../../private/tmp/metabase-20170726-52754-149xyg7/.brew_home/.config/yarn/global/node_modules/mkdirp/bin/cmd.js' -> '/usr/local/Cellar/node/8.2.1/bin/mkdirp'".
```

DomT4 replied:
>  which is imho harmful (nothing should manipulate another formulas cellar).

Agreed. On a practical level, also impossible these days with the sandbox in place. Thanks for creating the issue around `yarn`, I'll go read how that discussion went & chuck in any thoughts I have there.

chrmoritz replied:
> also impossible these days with the sandbox in place

But the sandbox is only active when executing brew commands. A user with the current yarn formula executing `yarn global add webpack` will create a symlink in `HOMEBREW_PREFIX/Cellar/node/8.2.1/bin/webpack`.

DomT4 replied:
Oh, yes, of course. Apologies; thought you were referencing that within the boundaries of Homebrew formulae.

chrmoritz commented:
I've opened #16083 to further discuss the yarn global installation issues.

I can confirm that the 2nd commit can be reverted, when you apply the 2nd fix from #16083 and the `fsevents` native addon gets still compiled successfully.

chrmoritz commented:
Looks like a compiled `fsevents` is not needed to run the metabase and skipping it like in the other formulae with are build with yarn is totally fine.

It had turned out, that the build failure I've observed earlier today was caused by a temporary outtake of the registry leiningen had to dowload packages from (which was my first guess) and not by the skipped fsevents module. (It was totally random, that the registry resolved it's problems after I've fixed the fsevents installation). 

I've reverted this PR back to the simple HEAD build fix (and leave the yarn issue to further discussion inside #16083).

ilovezfs commented:
Thanks @chrmoritz!
